pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/EtcdMetadataStore.java

```diff
@@ -0,0 +1,426 @@
+                            OpDelete del = op.asDelete();
+                            DeleteResponse dr = txnResponse.getDeleteResponses().get(deletedIdx++);
+                            if (dr.getDeleted() == 0) {
+                                del.getFuture().completeExceptionally(new MetadataStoreException.NotFoundException());
+                            } else {
+                                del.getFuture().complete(null);
+                            }
+                            break;
+                        }
+                        case GET_CHILDREN: {
+                            OpGetChildren getChildren = op.asGetChildren();
+                            GetResponse gr = txnResponse.getGetResponses().get(getIdx++);
+                            String basePath = getChildren.getPath() + "/";
+
+                            Set<String> children = gr.getKvs().stream()
+                                    .map(kv -> kv.getKey().toString(StandardCharsets.UTF_8))
+                                    .map(p -> p.replace(basePath, ""))
+                                    // Only return first-level children
+                                    .map(k -> k.split("/", 2)[0])
+                                    .collect(Collectors.toCollection(TreeSet::new));
+
+                            getChildren.getFuture().complete(new ArrayList<>(children));
+                        }
+                    }
+                }
+            }).exceptionally(ex -> {
+                Throwable cause = ex.getCause();
+                if (cause instanceof ExecutionException || cause instanceof CompletionException) {
+                    cause = cause.getCause();
+                }
+                if (ops.size() > 1 && cause instanceof StatusRuntimeException) {
+                    Status.Code code = ((StatusRuntimeException) cause).getStatus().getCode();
+                    if (
+                            code == Status.Code.INVALID_ARGUMENT /* This could be caused by having repeated keys
+                                    in the batch, retry individually */
+                                    ||
+                                    code
+                                            == Status.Code.RESOURCE_EXHAUSTED /* We might have exceeded the max-frame
+                                             size for the response */
+                    ) {
+                        ops.forEach(o -> batchOperation(Collections.singletonList(o)));
+                    }
+                } else {
+                    log.warn("Failed to commit: {}", cause.getMessage());
+                    ops.forEach(o -> o.getFuture().completeExceptionally(ex));
+                }
+                return null;
+            });
+        } catch (Throwable t) {
+            log.warn("Error in committing batch: {}", t.getMessage());
+        }
+    }
+
+    private synchronized CompletableFuture<Void> createLease(boolean retryOnFailure) {
```

Comment:
Does `retryOnFailure` is to retry in the background but the caller will get an exception?

Reply:
No, this is after we created the lease the first time, so that we can re-create it in background. If we fail, we notify the session listener to change the current state and we keep retrying in background.